Dockerfile

```diff
@@ -1,17 +1,25 @@
 FROM golang:alpine
 
 # Create application directory
-RUN mkdir /app
-ADD . /app/
-WORKDIR /app
+RUN mkdir /src
+ADD . /src/
+WORKDIR /src
+ENV CGO_ENABLED=0 GOOS=linux GOARCH=amd64
 
 # Build the application
-RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go mod download
-RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -o run .
+COPY go.* .
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 \
+    --mount=type=cache,target=/go/pkg/mod \
+    go mod graph | awk '{if ($1 !~ "@") print $2}' | xargs -r go get
```

Comment:
Could you elaborate how that is in any way better then `go mod download`? My understand is, that both commands will end up downloading everything, but `go mod download` seems to be much cleaner, no?

Reply:
I found this from this repo here: https://github.com/montanaflynn/golang-docker-cache 
From what I understand, the `go get` approach also compiles the dependencies. This should save time when you rebuild the image, as the dependencies won't need to be recompiled. Though I did a more thorough testing for this and found some rather interesting result which isn't quite what I expected.

1. Originally building is painfully slow, taking 140 seconds to build and the image is over 1 GB in size. Rebuilding isn't much faster without caching, taking 97 seconds instead.

2. Using `go mod download` with cache reduces the image size to ~310 MB but doesn't reduce initial the build time. Rebuilding takes only 24 seconds though.

3. My approach only reduces the image size to ~330MB, but reduce the initial build time by to 85 seconds. Rebuilding takes slightly more time though, with roughly 26 seconds which is probably within the margins of error.

So I feel like caching significantly reduces the rebuild time, while using the `go get` approach reduces the amount of dependencies required to be downloaded compared to `go mod download`.